src/xview/moxftp-2.2/Wcl-2.4/Xp/Xp.man
.COMMENT SccsID: @(#) Xp.man 1.1 92/03/18 11:17:08
.TH "Xp" 1 "1 March 1992"
.SH NAME
Xp \- Xt and Athena Public Library
.SH SYNOPSIS
.ta 1.5i 2.5i
.nf
.ft B
#include <X11/Intrinsic.h>
#include <X11/Wc/WcCreate.h>
#include <X11/Xp/Xp.h>

/* All Wcl applications should provide at least the Wcl options:
*/
static XrmOptionDescRec options[] = {
    ... application specific command line options ...
    WCL_XRM_OPTIONS
};

main ( argc, argv )
    int    argc;
    char*  argv[];
{
    /*  -- Intialize Toolkit creating the application shell
    */
    Widget appShell = XtInitialize (
        argv[0], WcAppClass( argc, argv ),      /* app name & class  */
        options, XtNumber(options),             /* argv option descr */
        &argc, argv
    );
    XtAppContext app = XtWidgetToApplicationContext(appShell);

    /*  -- Register application specific callbacks and widget classes
    */
    RegisterApplication ( app );

    /*  -- Register all widget classes and constructors
    */
    XpRegisterAthena( app );

    /*  -- Create widget tree below toplevel shell using Xrm database
    */
    if ( WcWidgetCreation ( appShell ) )
        exit(1);

    /*  -- Realize the widget tree
    */
    XtRealizeWidget ( appShell );

    /*  -- and finally, enter the main application loop
    */
    XtMainLoop ( );
}
.fi
.ft R

.SH LINK BY
.RS
.nf
.ft B
example% ld -o prog prog.o -lXp -lXaw -lWc <dynLinkLib> -lXmu -lXt -lX11 -lm
.ft R
.RE
.LP 
On SunOS or SVR4 machines, the <dynLinkLib> is named -ldl.
Other architectures cannot use the dynamic linking capabilites of
Wcl, and therefore do not need any dynamic link library.
.SH DESCRIPTION
.LP
.I Xp
is intended to be used with
.I Wcl
and the Athena widget set.  Xp provides a procedure for registering all
Athena widgets, and includes the XpTable widget.
.SH REGISTRATION PROCEDURES
.LP
.I Wcl
must be told the names of all widgets and widget constructors before
it can create widgets as specified in the resource file of client
applications.
.I Xp
provides a registration procedure for registering all Athena widgets.
For backward compatibility with previous releases of Wcl, this
registration procedure has three names: AriRegisterAthena(), 
XpRegisterAll(), and XpRegisterAthena().  These procedures
are exactly equivalent.
.TP
.B void XpRegisterAthena( XtAppContext )
This procedure registers all of the Athena widget class pointer names,
widget class names, and widget constructor function names with Wcl.
It also registers all Xp converters and action procedures with Xt,
and all Xp callback procedures with Wcl.
.SH CONVERTERS
.LP
.I Xp
does not provide any converters itself, but it does attempt to ensure
that WcCvtStringToWidget is used rather than XmuNewStringToWidget.
.SH "SEE ALSO"
.BI XpTable (3),
.BI Ari (1),	
.BI Wcl (3),
.BI X (1)
.SH AUTHOR
David E. Smyth (David.Smyth@sniap.mchp.sni.de) at Siemens Nixdorf
Informationssysteme AG, Munich Germany.


src/NeWS
../NeWS

src/xview/moxftp-2.2/Wcl-2.4/Xp/Xp.c
#include <X11/Wc/COPY>

/*
* SCCS_data: @(#) Xp.c	1.2 92/10/28 07:53:56
*
*     This module contains registration routine for all Athena
*     widget constructors and classes.  
*
* Module_interface_summary: 
*
*     void XpRegisterAthena  ( XtAppContext app )
*     void XpRegisterAll     ( XtAppContext app )
*     void AriRegisterAthena ( XtAppContext app )
*
*******************************************************************************
*/

#include <X11/Xatom.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>

#include <X11/Wc/WcCreateP.h>
#include <X11/Xp/Xp.h>

#ifdef DEBUG
#include <X11/Xp/XpAthenaP.h>
#else
#include <X11/Xp/XpAthena.h>
#endif

#if defined(XtSpecificationRelease) && XtSpecificationRelease > 4
#define R5
#endif

/* --  Register all Athena widgets, converters, callbacks, actions.
*******************************************************************************
*/

void AriRegisterAthena ( app )
    XtAppContext app;
{
    XpRegisterAthena(app);
}

void XpRegisterAll ( app )
    XtAppContext app;
{
    XpRegisterAthena(app);
}

void XpRegisterAthena ( app )
    XtAppContext app;
{
    ONCE_PER_XtAppContext( app );

/* -- Force Athena to load XmuNewCvtStringToWidget so
 *    WcCvtStringToWidget stays in effect when loaded by WcInitialize()
 */
    XtInitializeWidgetClass( toggleWidgetClass );
    XtInitializeWidgetClass( formWidgetClass );
    
#define RCP( name, class ) WcRegisterClassPtr  ( app, name, class );
#define RCO( name, func )  WcRegisterConstructor(app, name, func  )

    /* -- register all Xp widget classes */
    RCP("XpTable",			xpTableWidgetClass	);
    RCP("xpTableWidgetClass",		xpTableWidgetClass	);

    /* -- register all Athena widget classes */
    /* Simple Widgets (Chapt 3) */
    RCP("Command",			commandWidgetClass	);
    RCP("commandWidgetClass",		commandWidgetClass	);
    RCP("Core",				coreWidgetClass		);
    RCP("coreWidgetClass",		coreWidgetClass		);
    RCP("Grip",				gripWidgetClass		);
    RCP("gripWidgetClass",		gripWidgetClass		);
    RCP("Label",			labelWidgetClass	);
    RCP("labelWidgetClass",		labelWidgetClass	);
    RCP("List",				listWidgetClass		);
    RCP("listWidgetClass",		listWidgetClass		);
    RCP("Object",			objectClass		);
    RCP("objectClass",			objectClass		);
#ifdef R5
    RCP("Panner",			pannerWidgetClass	);
    RCP("pannerWidgetClass",		pannerWidgetClass	);
    RCP("Porthole",			portholeWidgetClass	);
    RCP("portholeWidgetClass",		portholeWidgetClass	);
#endif
    RCP("Rect",				rectObjClass		);
    RCP("rectObjClass",			rectObjClass		);
#ifdef R5
    RCP("Repeater",			repeaterWidgetClass	);
    RCP("repeaterWidgetClass",		repeaterWidgetClass	);
#endif
    RCP("Scrollbar",			scrollbarWidgetClass	);
    RCP("scrollbarWidgetClass",		scrollbarWidgetClass	);
    RCP("Simple",			simpleWidgetClass	);
    RCP("simpleWidgetClass",		simpleWidgetClass	);
    RCP("StripChart",			stripChartWidgetClass	);
    RCP("stripChartWidgetClass",	stripChartWidgetClass	);
    RCP("Toggle",			toggleWidgetClass	);
    RCP("toggleWidgetClass",		toggleWidgetClass	);

    /* Menus (Chapt 4) */
    RCP("SimpleMenu",			simpleMenuWidgetClass	);
    RCP("simpleMenuWidgetClass",	simpleMenuWidgetClass	);
    RCO("XawCreateSimpleMenu",		XpCreateSimpleMenu	);
    RCP("SmeBSB",			smeBSBObjectClass	);
    RCP("smeBSBObjectClass",		smeBSBObjectClass	);
    RCP("SmeLine",			smeLineObjectClass	);
    RCP("smeLineObjectClass",		smeLineObjectClass	);
    RCP("Sme",				smeObjectClass		);
    RCP("smeObjectClass",		smeObjectClass		);
    RCP("MenuButton",			menuButtonWidgetClass	);
    RCP("menuButtonWidgetClass",	menuButtonWidgetClass	);

    /* Text Widgets (Chapt 5) */
    RCP("AsciiText",			asciiTextWidgetClass	); /* NB name */
    RCP("asciiTextWidgetClass",		asciiTextWidgetClass	);
    RCP("AsciiSrc",			asciiSrcObjectClass	);
    RCP("asciiSrcObjectClass",		asciiSrcObjectClass	);
    RCP("AsciiSink",			asciiSinkObjectClass	);
    RCP("asciiSinkObjectClass",		asciiSinkObjectClass	);
    RCP("TextSink",			textSinkObjectClass	);
    RCP("textSinkObjectClass",		textSinkObjectClass	);
    RCP("TextSrc",			textSrcObjectClass	);
    RCP("textSrcObjectClass",		textSrcObjectClass	);
    RCP("Text",				textWidgetClass		);
    RCP("textWidgetClass",		textWidgetClass		);

    /* Composite and Constraint Widgets (Chapt 6) */
    RCP("ApplicationShell",		applicationShellWidgetClass	);
    RCP("applicationShellWidgetClass",	applicationShellWidgetClass	);
    RCP("Box",				boxWidgetClass		);
    RCP("boxWidgetClass",		boxWidgetClass		);
    RCP("Composite",			compositeWidgetClass	);
    RCP("compositeWidgetClass",		compositeWidgetClass	);
    RCP("Constraint",			constraintWidgetClass	);
    RCP("constraintWidgetClass",	constraintWidgetClass	);
    RCP("Dialog",			dialogWidgetClass	);
    RCP("dialogWidgetClass",		dialogWidgetClass	);
    RCP("Form",				formWidgetClass		);
    RCP("formWidgetClass",		formWidgetClass		);
    RCP("OverrideShell",		overrideShellWidgetClass	);
    RCP("overrideShellWidgetClass",	overrideShellWidgetClass	);
    RCP("Paned",			panedWidgetClass	);
    RCP("panedWidgetClass",		panedWidgetClass	);
    RCP("Shell",			shellWidgetClass	);
    RCP("shellWidgetClass",		shellWidgetClass	);
    RCP("TopLevelShell",		topLevelShellWidgetClass	);
    RCP("topLevelShellWidgetClass",	topLevelShellWidgetClass	);
    RCP("TransientShell",		transientShellWidgetClass	);
    RCP("transientShellWidgetClass",	transientShellWidgetClass	);
#ifdef R5
    RCP("Tree",				treeWidgetClass		);
    RCP("treeWidgetClass",		treeWidgetClass		);
#endif
    RCP("VendorShell",			vendorShellWidgetClass	);
    RCP("vendorShellWidgetClass",	vendorShellWidgetClass	);
    RCP("Viewport",			viewportWidgetClass	);
    RCP("viewportWidgetClass",		viewportWidgetClass	);
    RCP("WmShell",			wmShellWidgetClass	);
    RCP("wmShellWidgetClass",		wmShellWidgetClass	);

    /* Other Interestng Widgets (not in ref manual) */
#ifndef R6
    RCP("ClockWidget",			clockWidgetClass	);
    RCP("clockWidgetClass",		clockWidgetClass	);
    RCP("LogoWidget",			logoWidgetClass		);
    RCP("logoWidgetClass",		logoWidgetClass		);
#endif
#if defined(R5)&&!defined(R6)
    RCP("Mailbox",			mailboxWidgetClass	);
    RCP("mailboxWidgetClass",		mailboxWidgetClass	);
#endif
}

/*
    -- Create SimpleMenu as pop-up child
*******************************************************************************
*/

Widget XpCreateSimpleMenu( pw, name, args, nargs )
    Widget      pw;     /* children's parent                            */
    String      name;   /* widget name to create                        */
    Arg        *args;   /* args for widget                              */
    Cardinal    nargs;  /* args count                                   */
{
  return XtCreatePopupShell(name, simpleMenuWidgetClass, pw, args, nargs);
}

/* backward compatibility:
*/
Widget WcCreateSimpleMenu( pw, name, args, nargs )
    Widget      pw;     /* children's parent                            */
    String      name;   /* widget name to create                        */
    Arg        *args;   /* args for widget                              */
    Cardinal    nargs;  /* args count                                   */
{
  return XtCreatePopupShell(name, simpleMenuWidgetClass, pw, args, nargs);
}

/*
    -- Append to end of XawText widget
*******************************************************************************
*/

int XpTextAppend( w, msg )
    Widget w;
    char*  msg;
{
    XawTextBlock textBlock;
    XawTextPosition textPos;

    textPos = XawTextGetInsertionPoint( w );
    textBlock.firstPos = 0;			/* start at ptr */
    textBlock.length = strlen( msg );
    textBlock.ptr = msg;
    textBlock.format = FMT8BIT;
    return XawTextReplace( w, textPos, textPos, &textBlock );
}
